### ISSUE TITLE
rr first_instruction selftests fail

### ISSUE THREAD

**Issue by pm215** [mozilla/rr#1536]
Running 'make check' on current rr git (commit d352b499ec399b) results in test failures:

```
1265/1268 Test #1251: string_instructions_break-32 ..............................   Passed    4.50 sec
1266/1268 Test #1252: string_instructions_break-32-no-syscallbuf ................   Passed    4.55 sec
1189: FAILED: expecting "function _start"
1189: exception: Timeout exceeded.
1189: <pexpect.spawn object at 0x2b64a5acbbd0>
1189: version: 3.1
1189: command: /home/petmay01/src/rr-obj64/bin/rr
1189: args: ['/home/petmay01/src/rr-obj64/bin/rr', '--suppress-environment-warnings', '--check-cached-mmaps', '--fatal-errors', 'replay', '-x', '/home/petmay01/src/rr/src/test/test_setup.gdb']
1189: searcher: <pexpect.searcher_re object at 0x2b64a5acbc50>
1189: buffer (last 100 chars): ' disass\r\nNo function contains program counter for selected frame.\r\n(gdb) '
1189: before (last 100 chars): ' disass\r\nNo function contains program counter for selected frame.\r\n(gdb) '
1189: after: <class 'pexpect.TIMEOUT'>
1189: match: None
1189: match_index: None
1189: exitstatus: None
1189: flag_eof: False
1189: pid: 6594
1189: child_fd: 6
1189: closed: False
1189: timeout: 100
1189: delimiter: <class 'pexpect.EOF'>
1189: logfile: <open file 'gdb_rr.log', mode 'w' at 0x2b64a5a769c0>
1189: logfile_read: None
1189: logfile_send: None
1189: maxread: 2000
1189: ignorecase: False
1189: searchwindowsize: None
1189: delaybeforesend: 0.05
1189: delayafterclose: 0.1
1189: delayafterterminate: 0.1
1189: Test 'first_instruction_32' FAILED: debug script failed
1189: Test first_instruction_32 failed, leaving behind /tmp/rr-test-first_instruction_32-vPCIeh2C7
1189: To replay the failed test, run
1189:   _RR_TRACE_DIR=/tmp/rr-test-first_instruction_32-vPCIeh2C7 rr replay
1267/1268 Test #1189: first_instruction-32 ......................................***Failed  Error regular expression found in output. Regex=[FAILED]100.57 sec
1190: FAILED: expecting "function _start"
1190: exception: Timeout exceeded.
1190: <pexpect.spawn object at 0x2b835b536bd0>
1190: version: 3.1
1190: command: /home/petmay01/src/rr-obj64/bin/rr
1190: args: ['/home/petmay01/src/rr-obj64/bin/rr', '--suppress-environment-warnings', '--check-cached-mmaps', '--fatal-errors', 'replay', '-x', '/home/petmay01/src/rr/src/test/test_setup.gdb']
1190: searcher: <pexpect.searcher_re object at 0x2b835b536c50>
1190: buffer (last 100 chars): ' disass\r\nNo function contains program counter for selected frame.\r\n(gdb) '
1190: before (last 100 chars): ' disass\r\nNo function contains program counter for selected frame.\r\n(gdb) '
1190: after: <class 'pexpect.TIMEOUT'>
1190: match: None
1190: match_index: None
1190: exitstatus: None
1190: flag_eof: False
1190: pid: 6677
1190: child_fd: 6
1190: closed: False
1190: timeout: 100
1190: delimiter: <class 'pexpect.EOF'>
1190: logfile: <open file 'gdb_rr.log', mode 'w' at 0x2b835b4e19c0>
1190: logfile_read: None
1190: logfile_send: None
1190: maxread: 2000
1190: ignorecase: False
1190: searchwindowsize: None
1190: delaybeforesend: 0.05
1190: delayafterclose: 0.1
1190: delayafterterminate: 0.1
1190: Test 'first_instruction_32' FAILED: debug script failed
1190: Test first_instruction_32 failed, leaving behind /tmp/rr-test-first_instruction_32-RNF8Rf56C
1190: To replay the failed test, run
1190:   _RR_TRACE_DIR=/tmp/rr-test-first_instruction_32-RNF8Rf56C rr replay
1268/1268 Test #1190: first_instruction-32-no-syscallbuf ........................***Failed  Error regular expression found in output. Regex=[FAILED]100.70 sec

99% tests passed, 3 tests failed out of 1268

Total Test time (real) = 688.14 sec

The following tests FAILED:
        1100 - string_instructions-32-no-syscallbuf (Failed)
        1189 - first_instruction-32 (Failed)
        1190 - first_instruction-32-no-syscallbuf (Failed)
Errors while running CTest
make[3]: *** [CMakeFiles/check] Error 8
```

This is on an Ubuntu trusty x86-64 host.

It looks like this is because the debugger gets control in /lib/ld-linux.so.2 somewhere with no debug info, which is somewhat earlier in execution than _start.

(Not sure about the string_instructions failure, there doesn't seem to be anything immediately obvious in the 'make check' output about that test.)

**Comment by pm215** [comment]
In a rerun of make check the string_instructions test passed; no idea why it failed first time around.

**Comment by rocallahan** [comment]
Odd. If you just start replay of `simple_32`, what do you get for `disassemble`?

**Comment by pm215** [comment]
Reading symbols from /lib/ld-linux.so.2...(no debugging symbols found)...done.
Loaded symbols for /lib/ld-linux.so.2
0xf777b0d0 in ?? () from /lib/ld-linux.so.2
(gdb) disas
No function contains program counter for selected frame.
(gdb) disas $pc,$pc+32
Dump of assembler code from 0xf777b0d0 to 0xf777b0f0:
=> 0xf777b0d0:  mov    %esp,%eax
   0xf777b0d2:  call   0xf777e7c0
   0xf777b0d7:  mov    %eax,%edi
   0xf777b0d9:  call   0xf777b0c0
   0xf777b0de:  add    $0x1ff22,%ebx
   0xf777b0e4:  mov    -0x104(%ebx),%eax
   0xf777b0ea:  pop    %edx
   0xf777b0eb:  lea    (%esp,%eax,4),%esp
   0xf777b0ee:  sub    %eax,%edx
End of assembler dump.

I'm not sure exactly where we are here, but my guess is we're in dl_main().

**Comment by rocallahan** [comment]
Can you find that code in objdump -d to be sure?

I think we might just have to disable that test in -32.

**Comment by pm215** [comment]
From objdump -d /lib/ld-linux.so.2, but not very informative because there's no debug info:

```
    10d0:       89 e0                   mov    %esp,%eax
    10d2:       e8 e9 36 00 00          call   47c0 <free@plt+0x3f70>
    10d7:       89 c7                   mov    %eax,%edi
    10d9:       e8 e2 ff ff ff          call   10c0 <free@plt+0x870>
    10de:       81 c3 22 ff 01 00       add    $0x1ff22,%ebx
    10e4:       8b 83 fc fe ff ff       mov    -0x104(%ebx),%eax
    10ea:       5a                      pop    %edx
    10eb:       8d 24 84                lea    (%esp,%eax,4),%esp
    10ee:       29 c2                   sub    %eax,%edx
    10f0:       52                      push   %edx
    10f1:       8b 83 40 00 00 00       mov    0x40(%ebx),%eax
```

I think I was wrong about this being dl_main.

It is, curiously, exactly 0x200 before _start (which is the specified ELF entry point for the dynamic linker). I'm now wondering if there's possibly a bug in how rr picks the starting instruction when it kicks things off?

**Comment by rocallahan** [comment]
We don't pick the instruction. We're supposed to stop at exactly whatever $eip is immediately after execve().

`rr dump` should show the $eip value we recorded immediately after exec. Does that match what you see in gdb (0xf777b0d0)?

**Comment by pm215** [comment]
rr dump has this event:
```
{
  global_time:14, event:`SYSCALL: execve' (state:EXITING_SYSCALL) tid:9808, ticks:0
 eax:0x0 ecx:0x0 edx:0x0 ebx:0x0 esp:0xffe65400 ebp:0x0 esi:0x0 edi:0x0 eip:0xf777b0d0 eflags:0x246 xcs:0x23 xss:0x2b xds:0x2b xes:0x2b xfs:0x0 xgs:0x0 orig_eax:0xb
}
```
which I guess implies that we were told that's the right EIP for exiting execve. And actually that is the ELF entry point for /lib/ld-linux.so.2. I was confused because the debug-symbols version of the library has a different entry point for some reason.

**Comment by pm215** [comment]
Aha. I've figured out what's happening here. Most of my confusion was caused by looking at the debug-symbols version of the 64-bit ld.so rather than the 32-bit version, which wasn't installed.

And that's actually what's going on here. If I install the 32-bit version of the ld.so debug symbols, then 'disas' says:
```
Dump of assembler code for function _start:
=> 0xf777b0d0 <+0>:     mov    %esp,%eax
   0xf777b0d2 <+2>:     call   0xf777e7c0 <_dl_start>
```
So the problem is just that the test case is relying on the system having debug information available for the dynamic linker, which isn't guaranteed to be the case. It's also not 32-bit specific -- I expect that if I didn't have the debug info for 64 bit installed then the test case would fail there too.

A couple of possibilities spring to mind for versions of this test that don't require debug info for the dynamic linker:
 (1) use a statically linked test binary
 (2) identify the dynamic linker being used for the text executable and then fish the ELF entry point out of it (via objdump or otherwise) and then check that the entry point PC is the one you get as first insn of replay
